linear_algebra/tensor_product.lean

```diff
@@ -0,0 +1,378 @@
+/-
+Copyright (c) 2018 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Tensor product of modules over commutative rings.
+
+-/
+
+import group_theory.free_abelian_group
+import linear_algebra.linear_map_module
+
+universes u v w u₁ v₁
+
+variables {R : Type u} [comm_ring R]
+variables (M : Type v) (N : Type w) (P : Type u₁) (Q : Type v₁)
+variables [module R M] [module R N] [module R P] [module R Q]
+include R
+
+section bilinear
+
+variables {M N P Q}
+structure is_bilinear_map {M N P}
+  [module R M] [module R N] [module R P]
+  (f : M → N → P) : Prop :=
+(add_pair : ∀ x y z, f (x + y) z = f x z + f y z)
+(pair_add : ∀ x y z, f x (y + z) = f x y + f x z)
+(smul_pair : ∀ r x y, f (r • x) y = r • f x y)
+(pair_smul : ∀ r x y, f x (r • y) = r • f x y)
+
+variables {f : M → N → P} (hf : is_bilinear_map f)
+include hf
+
+theorem is_bilinear_map.zero_pair : ∀ y, f 0 y = 0 :=
+λ y, calc f 0 y
+        = f (0 + 0) y - f 0 y : by rw [hf.add_pair 0 0 y]; simp
+    ... = 0 : by simp
+
+theorem is_bilinear_map.pair_zero : ∀ x, f x 0 = 0 :=
+λ x, calc f x 0
+        = f x (0 + 0) - f x 0 : by rw [hf.pair_add x 0 0]; simp
+    ... = 0 : by simp
+
+theorem is_bilinear_map.neg_pair : ∀ x y, f (-x) y = -f x y :=
+λ x y, by convert hf.smul_pair (-1) x y; simp
+
+theorem is_bilinear_map.pair_neg : ∀ x y, f x (-y) = -f x y :=
+λ x y, by convert hf.pair_smul (-1) x y; simp
+
+theorem is_bilinear_map.linear_pair (y : N) : is_linear_map (λ x, f x y) :=
+{ add  := λ m n, hf.add_pair m n y,
+  smul := λ r m, hf.smul_pair r m y }
+
+theorem is_bilinear_map.pair_linear (x : M) : is_linear_map (λ y, f x y) :=
+{ add  := λ m n, hf.pair_add x m n,
+  smul := λ r m, hf.pair_smul r x m }
+
+theorem is_bilinear_map.comm : is_bilinear_map (λ y x, f x y) :=
+by cases hf; constructor; intros; solve_by_elim
+
+variables {g : P → Q} (hg : is_linear_map g)
+include hg
+
+theorem is_bilinear_map.comp : is_bilinear_map (λ x y, g (f x y)) :=
+{ add_pair  := λ x y z, by rw [hf.add_pair, hg.add],
+  pair_add  := λ x y z, by rw [hf.pair_add, hg.add],
+  smul_pair := λ r x y, by rw [hf.smul_pair, hg.smul],
+  pair_smul := λ r x y, by rw [hf.pair_smul, hg.smul] }
+
+end bilinear
+
+namespace tensor_product
+
+def relators : set (free_abelian_group (M × N)) :=
+add_group.closure { x : free_abelian_group (M × N) |
+  (∃ (m1 m2 : M) (n : N), x = (m1, n) + (m2, n) - (m1 + m2, n)) ∨
+  (∃ (m : M) (n1 n2 : N), x = (m, n1) + (m, n2) - (m, n1 + n2)) ∨
+  (∃ (r : R) (m : M) (n : N), x = (r • m, n) - (m, r • n)) }
+

```

Comment:
Can you add these three cases as separate lemmas? Later it looks like you use a lot of `in_closure` with `inl`, `inr` combination.

Reply:
Actually I think this should be an inductive predicate, or possibly an inductive binary relation with `eqv_gen` applied instead of `add_group.closure`.

Reply:
I don't think that makes much of a difference since they are not part of the interface, and the corresponding interface theorems are already there (add_of, of_add, smul_of, of_smul).